src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -0,0 +1,423 @@
+            }
+        }
+
+        private static BoundDecisionDag MakeDecisionDag(ImmutableArray<PartialCaseDecision> cases)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    internal class PartialCaseDecision
+    {
+        public readonly int Index;
+        public readonly ImmutableArray<BoundDagDecision> Decisions;
+        public readonly ImmutableArray<(BoundExpression, BoundDagTemp)> Bindings;
+        public readonly BoundExpression WhereClause;
+        public readonly LabelSymbol CaseLabel;
+        public PartialCaseDecision(
+            int Index,
+            ImmutableArray<BoundDagDecision> Decisions,
+            ImmutableArray<(BoundExpression, BoundDagTemp)> Bindings,
+            BoundExpression WhereClause,
+            LabelSymbol CaseLabel)
+        {
+            this.Index = Index;
+            this.Decisions = Decisions;
+            this.Bindings = Bindings;
+            this.WhereClause = WhereClause;
+            this.CaseLabel = CaseLabel;
+        }
+    }
+
+    partial class BoundDagEvaluation
+    {
+        public override bool Equals(object obj) => obj is BoundDagEvaluation other && this.Equals(other);
+        public bool Equals(BoundDagEvaluation other)
+        {
+            return other != (object)null && this.Kind == other.Kind && this.Input.Equals(other.Input) && this.Symbol == other.Symbol;
+        }
+        private Symbol Symbol
+        {
+            get
+            {
+                switch (this)
+                {
+                    case BoundDagFieldEvaluation e: return e.Field;
+                    case BoundDagPropertyEvaluation e: return e.Property;
+                    case BoundDagTypeEvaluation e: return e.Type;
+                    case BoundDagDeconstructEvaluation e: return e.DeconstructMethod;
+                    default: throw ExceptionUtilities.UnexpectedValue(this.Kind);
+                }
+            }
+        }
+        public override int GetHashCode()
+        {
+            return this.Input.GetHashCode() ^ (this.Symbol?.GetHashCode() ?? 0);
+        }
+        public static bool operator ==(BoundDagEvaluation left, BoundDagEvaluation right)
+        {
+            return (left == (object)null) ? right == (object)null : left.Equals(right);
+        }
+        public static bool operator !=(BoundDagEvaluation left, BoundDagEvaluation right)
+        {
+            return !left.Equals(right);
+        }
+    }
+
+    partial class BoundDagTemp
+    {
+        public override bool Equals(object obj) => obj is BoundDagTemp other && this.Equals(other);
+        public bool Equals(BoundDagTemp other)
+        {
+            return other != (object)null && this.Type == other.Type && object.Equals(this.Source, other.Source) && this.Index == other.Index;
+        }
+        public override int GetHashCode()
+        {
+            return this.Type.GetHashCode() ^ (this.Source?.GetHashCode() ?? 0) ^ this.Index;
```

Comment:
>this.Type.GetHashCode() ^ (this.Source?.GetHashCode() ?? 0) ^ this.Index [](start = 19, length = 72)

Please use helpers to combine hash codes.

Reply:
Resolved in [https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec](https://github.com/dotnet/roslyn/pull/23357/commits/758fd76932a6d9bcb640504c8fdef57fb5adecec) of [https://github.com/dotnet/roslyn/pull/23357](https://github.com/dotnet/roslyn/pull/23357)

---
In reply to: [156559822](https://github.com/dotnet/roslyn/pull/23209#discussion_r156559822) [](ancestors = 156559822)